### PR TITLE
fix: in the stream mode, request cannot be canceled

### PR DESCRIPTION
**Pull request by lh900519**
### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [ ] I have added the required tests to prove the fix/feature I'm adding
- [ ] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [ ] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

**Comment by kuhnroyal** [comment]
Thanks for your PR!

Do you have any additional information for the root problem?
The fix currently won't work when thinking about parallel requests, the closing of the HTTP client may kill other requests.
Please provide a test that reproduces the problem.

**Comment by lh900519** [comment]
```dart
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:dio/dio.dart';

void main() async {
  final dio = Dio();
  clientStream(dio, '1');
  clientStream(dio, '2');
}

Future clientStream(Dio dio, String name) async {
  try {
    final token = CancelToken();
    Timer(const Duration(seconds: 3), () {
      token.cancel("----Canceled");
    });
    const url = "https://hacker-news.firebaseio.com/v0/topstories.json";
    Response<ResponseBody> rs = await dio.get<ResponseBody>(
      url,
      options: Options(
        headers: {
          "Authorization": 'vhdrjb token"',
          "Accept": "text/event-stream",
          "Cache-Control": "no-cache",
        },
        responseType: ResponseType.stream,
      ), // set responseType to `stream`
      cancelToken: token,
    );

    rs.data?.stream
        .transform(unit8Transformer)
        .transform(const Utf8Decoder())
        .transform(const LineSplitter())
        .listen((event) {
      if (event.startsWith("event:")) {
        print("$name $event isCancelled:${token.isCancelled}");
      }
      if (event.startsWith("data:")) {
        print(
            "$name ${event.substring(0, 5)} isCancelled:${token.isCancelled}");
      }
    });
  } catch (e) {
    print('error $e');
  }
}

StreamTransformer<Uint8List, List<int>> unit8Transformer =
    StreamTransformer.fromHandlers(
  handleData: (data, sink) {
    sink.add(List<int>.from(data));
  },
);

```

Run the code with `dart test.dart`
The request was canceled using CancelToken after 3 seconds, but the terminal did not stop and continued to output.

This PR does not reuse _cachedHttpClient for stream type requests, but creates a new httpclient, and closes the newly created httpclient when using CancelToken. The processing of stream types here is the same as in 5.1.2.

File: dio/lib/src/adapters/io_adapter.dart [hunk 1]
```diff
@@ -227,7 +238,13 @@ class IOHttpClientAdapter implements HttpClientAdapter {
     );
   }
 
-  HttpClient _configHttpClient(Duration? connectionTimeout) {
+  HttpClient _configHttpClient(Duration? connectionTimeout,
+      {bool newClient = false}) {
```
Comment (AlexV525) [inline dio/lib/src/adapters/io_adapter.dart:242]: These are not well-formatted (with trailing commas).

File: dio/lib/src/adapters/io_adapter.dart [hunk 2]
```diff
@@ -227,7 +238,13 @@ class IOHttpClientAdapter implements HttpClientAdapter {
...
+    if (newClient) {
+      final client = _createHttpClient();
```
Comment (AlexV525) [inline dio/lib/src/adapters/io_adapter.dart:244]: If we create new clients here, what about the closing after the client has been used?

File: dio/lib/src/adapters/io_adapter.dart (hunk 1 above)
Comment (AlexV525) [inline dio/lib/src/adapters/io_adapter.dart:242]: I'd prefer `forceNotCachedClient` for better understanding, or someone else can have better namings.

**Review by AlexV525 — CHANGES_REQUESTED** [review]
If you have a reproducible example, please make it a test, we need tests to avoid further breakages.

**Comment by kuhnroyal** [comment]
I am also wondering if there is an upstream issue for this problem. We `abort()` the request correctly, why does this not work?

**Comment by kuhnroyal** [comment]
Maybe this is the source of the problem? https://github.com/dart-lang/sdk/issues/51267

**Comment by AlexV525** [comment]
Converting this to a draft until we get progress with the implementation and tests.

**Comment by AlexV525** [comment]
Should be covered by #2032.

**Pull request closed without merging**